text/2019-03-27-follower-request-snapshot.md

```diff
@@ -0,0 +1,78 @@
+# Follower/Learner request snapshot actively
+
+## Summary
+
+In current design, a follower can only wait for the leader to send a snapshot
+iff the leader founds out the follower cannot catch up to its index. This RFC
+proposes to let a follower able to request snapshot from leader actively.
+
+## Motivation
+
+For the TiFlash project, which synchronizes data via learner's raft log, the
+region meta data is stored in memory and flushed to disk periodically, chances
+are that the region meta is lost during the process. In the other hand, we
+also need a solution to restore broken data written to TiFlash since TiFlash
+only owns one copy. Provided that the learner is able to request snapshot
+actively, we can help restore broken data, and the restore process will be
+faster and easier once TiFlash provides consistency check.
+
+## Detailed design
+
+### raft-rs
+
+We should add a new API for raft-rs to request a snapshot of a region via grpc.
+
+The request snapshot message is piggybacked by `reject` and `reject_hint`, it
```

Comment:
It may be hard to tell whether it's requesting snapshot actively or suffering data loss. In the later case, it has chance that there is no logs but committed logs are overwritten silently.